### PR TITLE
[ready] Add matrix_power

### PR DESCRIPTION
**Pull request by vishwakftw**
1. Tests added
2. Doc string added

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: This is blocked by https://github.com/pytorch/pytorch/pull/9949 .

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (fmassa) [inline aten/src/ATen/native/LinearAlgebra.cpp]: should this be `matrix_power` instead?

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (fmassa) [inline aten/src/ATen/native/LinearAlgebra.cpp]: could you add some comments explaining the reasoning, similar to what is done in the numpy version?

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Oops, sorry. Will fix it.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Sure, I will do that.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: I am confused why this is not causing the derivative to be computed. The derivative is defined for the inverse (in `tools/autograd/derivatives.yaml`, and hence this should cause an autodiff based derivative for `n < 0`.

Comment (ezyang) [reply aten/src/ATen/native/LinearAlgebra.cpp]: The problem is you're calling `at::native::inverse`, when you should call `at::inverse`. There is a long and sordid story about this; see https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#can-it-handle-being-passed-variables

Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Thank you @ezyang !

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: The derivative in this case is supposed to be a matrix of 0s . Is there a way to register the gradient here?

Comment (ezyang) [reply aten/src/ATen/native/LinearAlgebra.cpp]: I don't see any data dependence on a, so shouldn't this be the case automatically?

Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Unfortunately doesn't seem so, which is why I was puzzled.

Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Specifically, this is the error:
```
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

**Comment by vishwakftw** [comment]
This is ready for review.

**Comment by vishwakftw** [comment]
@fmassa Could you review this please?

**Comment by yf225** [comment]
@fmassa Ping :)

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (soumith) [inline aten/src/ATen/native/LinearAlgebra.cpp]: `a.clone().copy_` is two copies, when you can just get by with 1 copy. What you want is `at::empty_like(a)` maybe?

**Review by soumith — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Oh, yes. Thank you. I will fix it right away.

**Review by soumith — APPROVED** [review]

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Based on the test failure, I think doing this will detach `a` from the autograd graph. Is there any other way to keep `a` linked to the graph?

**Review by fmassa — COMMENTED** [review]
Apart from the issue in the `n==0` case, everything else looks good to me

**Comment by vishwakftw** [comment]
Is this good to go?

**Comment by yf225** [comment]
@vishwakftw I just retriggered the CI to see if the ASAN test failure is consistent, if so we probably should look at how to fix it or skip it.

**Comment by vishwakftw** [comment]
Build failure seems to be unrelated to the PR.

**Comment by vishwakftw** [comment]
@yf225 is this good to go? Sorry about the reminder.

File: test/test_torch.py
Comment (yf225) [inline test/test_torch.py]: @vishwakftw Do you mind moving this line to the top of the file? It seems that our internal tests don't support the import here very well :(

Comment (vishwakftw) [reply test/test_torch.py]: Sure, will do.

**Comment by vishwakftw** [comment]
@yf225 could you try importing now? I am sorry if you are busy with other tasks - thought I should send a reminder just in case.

**Comment by vishwakftw** [comment]
@yf225 is this good to go?

**Comment by yf225** [comment]
@vishwakftw Apologies for the delay - there are some internal dependency errors that I am still trying to figure out how to fix. I will have an update here as soon as possible.

**Comment by vishwakftw** [comment]
No worries @yf225 . Thank you for helping out.

**Comment by yf225** [comment]
@vishwakftw Sorry for the delay on this - I think your original idea of putting `from test_autograd import random_fullrank_matrix_distinct_singular_value` in `_test_matrix_power` is actually the correct one, as it prevents other tests that depend on `test_torch.py` to also depend on `test_autograd.py`. Do you mind reverting the commit for this (`dcf13e5`) and I can push it through internally? Thanks a lot and sorry for the confusion on this one.

**Comment by vishwakftw** [comment]
No problem @yf225. I will revert the previous commit right away. Thank you again for helping out with this.

**Comment by vishwakftw** [comment]
@yf225 did the internal tests pass?

**Pull request closed without merging**
